src/DebuggerData.cpp

```diff
@@ -275,7 +274,7 @@ bool Breakpoints::inCurrentSlot(const Breakpoint& bp)
 void Breakpoints::insertBreakpoint(Breakpoint& bp)
 {
 	auto it = ranges::upper_bound(breakpoints, bp.range,
-	                              [](auto& bp1, auto& bp2) { return bp1.start < bp2.start; },
+	                              [](auto& range1, auto& range2) { return range1 && range2 ? range1->start < range2->start : true; },
```

Comment:
Hmm, this is more complicated than I thought. The requirement on the comparison function is that it models a "strict weak ordering relation". For detail see: https://en.cppreference.com/w/cpp/named_req/Compare. This implementation doesn't meet those requirements. (At work I've seen some, hard-to-debug, cases caused by invalid comparsion functions.)

One of the properties of such a relation is that it's irreflexive (an element cannot be smaller than itself). In other words `comp(a, a)` must always be `false`. But with this implementation if both `range1` and `range2` point to (the same) empty range then the result is `true`.

But changing `true` to `false` in this implementation is not sufficient. Though it's more complicated to see why. I'll demonstrate with an example: Let `x = std::nullopt`,   `y = AddressRange{1}`,   `z = AddressRange{2}`. Then according to this comparison function (with true changed to false) we have: `comp(x, y) == false` and `comp(y, x) == false`. In other words: `x` and `y` are equivalent elements. And also`comp(x, z) == false` and `comp(z, x) == false`. So also `x` and `z` are equivalent elements. But if `x` and `y` are equivalent, and `x` and `z` are equivalent then (because of transitivity) also `y` and `z` must be equivalent. But this contradicts with `comp(y, z) == true`.

One possible fix is like this (transform to a `int` where -1 represents an absent range):
`auto it = upper_bound(breakpoints, bp.range, {}, [](auto& bp) -> int { return bp.range ? bp.range->start : -1; });`
This is OK because comparison on `int` is strictly ordered.

A more general (but more complication) alternative would be to first transform to a tuple with the first field in the tuple a boolean (range is absent or not) and the second field the start address (or a dummy value if absent). And then do a lexicographical comparison on this tuple (this is the default comparison function for tuples).

Reply:
So we are moving away from comparisons and going to projections, which is funny since we tried to get away from `-1` in the first place and adopt a more accurate presentation of the data.

Reply:
It's not about projections vs comparison. It's about the current comparison function that is mathematically wrong.

Technically everything that projections can do, and more, can also be done via comparisons. So comparison are strictly more powerful. Nevertheless I often find projections easier: the left- and the right-hand-side are automatically symmetrical. And you can often rely on a standard comparison function (typically the default operator<) which always models a mathematically correct strict (weak) ordering relation.

But your remark did give me an idea (because I agree the `-1` is not pretty). Earlier I already said that the more general solution is to project onto a tuple(bool, value). But in fact such a tuple is semantically the same as a std::optional. With this idea the code becomes:
`auto it = upper_bound(breakpoints, bp.range, {}, [](auto& bp) { return bp.range ? bp.range->start : std::nullopt; });`
... ah maybe that is what you meant with your remark?